In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score , confusion_matrix

In [2]:
#Read
df = pd.read_csv('news.csv')

C:\Users\om\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3057: DtypeWarning: Columns (24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
#removing above warnings
import warnings
warnings.filterwarnings('ignore')

In [4]:
df = pd.read_csv('news.csv')

In [5]:
df.shape

(7795, 141)

In [6]:
df.head()

Unnamed: 0                                              title  \
0       8476                       You Can Smell Hillary’s Fear   
1      10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2       3608        Kerry to go to Paris in gesture of sympathy   
3      10142  Bernie supporters on Twitter erupt in anger ag...   
4        875   The Battle of New York: Why This Primary Matters   

                                                text label Unnamed: 4  \
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE        NaN   
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE        NaN   
2  U.S. Secretary of State John F. Kerry said Mon...  REAL        NaN   
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE        NaN   
4  It's primary day in New York and front-runners...  REAL        NaN   

  Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 131  \
0        NaN        NaN        NaN        NaN        NaN  ...          NaN   
1        NaN        NaN        NaN        NaN        NaN  ...          NaN   
2        NaN        NaN        NaN        NaN        NaN  ...          NaN   
3        NaN        NaN        NaN        NaN        NaN  ...          NaN   
4        NaN        NaN        NaN        NaN        NaN  ...          NaN   

  Unnamed: 132 Unnamed: 133 Unnamed: 134 Unnamed: 135 Unnamed: 136  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

  Unnamed: 137 Unnamed: 138 Unnamed: 139 Unnamed: 140  
0          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN  

[5 rows x 141 columns]

In [7]:
#get labels
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [8]:
#split dataset for train and validate or test
x_train , x_test , y_train , y_test = train_test_split(df['text'] , labels , test_size = 0.2 , random_state = 0)

In [9]:
#initialize a tfidf vectorizer
tfidf_vectorizer = TfidfVectorizer(stop_words = 'english' , max_df = 0.7)


#fit and transform vectorizer on train set and transform test set
tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test  = tfidf_vectorizer.transform(x_test)

ValueError: np.nan is an invalid document, expected byte or unicode string.

In [10]:
#for removing the above error 
x_train.fillna(' ',inplace=True)
x_test.fillna(' ',inplace=True)
y_train.fillna(' ',inplace=True)
y_test.fillna(' ',inplace=True)

In [11]:
#initialize a tfidf vectorizer
tfidf_vectorizer = TfidfVectorizer(stop_words = 'english' , max_df = 0.7)


#fit and transform vectorizer on train set and transform test set
tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test  = tfidf_vectorizer.transform(x_test)

In [12]:
#initialize a PassiveAggressiveClassifier
pac = PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train , y_train)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

In [13]:
#predict on test set
y_pre = pac.predict(tfidf_test)

In [14]:
#calculate accuracy score
accuracy = accuracy_score(y_test , y_pre)
print(f'Accuracy : {round(accuracy*100,2)}%')

Accuracy : 77.68%


In [15]:
#build confusion matrix
confusion_matrix(y_test , y_pre , labels = ['FAKE','REAL'])

array([[588,  31],
       [ 46, 616]], dtype=int64)

So , we have 588 true positives , 31 false negatives , 616 true negatives , 46 false positives .